### ISSUE TITLE
no-relative-parent-imports handling of "internal" imports seems broken

### ISSUE THREAD

**Issue by jonathanj** [benmosher/eslint-plugin-import#1477]
I have a project that aliases `./src/app` as `app`, so project imports look like `import Foo from 'app/x/y`.

When trying to use the `no-relative-parent-imports` rule (as a warning) I am seeing the linter emit a warning in places I don't expect. For example in the file `src/app/components/one/index.js` an import such as `import Two from 'app/components/two` produces this (unexpected) warning.

Reading [the source for the rule](https://github.com/benmosher/eslint-plugin-import/blob/1a90a2083b3da9e384eea7ba2f1652619896083a/src/rules/no-relative-parent-imports.js) and applying some debugging, it seems like my import type is being detected as "internal" (which seems reasonable, it's not explicitly relative nor does it refer to an import from node_modules) but the subsequent handling of internal imports seems broken to me.

The code goes on to resolve my import to an absolute path and then creates a relative path from the directory of the source file being linted to the that absolute path, and uses *that* to determine the import type with which to conditionally emit a warning. In code:

```
> const relDepPath = relative('/Users/jonathan/Coding/project/src/app/components/one', '/Users/jonathan/Coding/project/src/app/components/two')
> relDepPath
'../two'
> importType(relDepPath, context)
'../parent'
```

The part that I'm confused about is that in the case of "internal" imports (e.g. I'm importing project code from project code) the relative relationship between the current file and the import target is extremely likely to require ascending the directory structure.

There is an escape hatch for "external" imports which I think is also the right way to handle "internal" imports, if the import target is specified in another way (say `../two`) then it will no longer be interpreted as an "internal" import and will be subject to the normal linting rule behaviour.

**Comment by ljharb** [comment]
That's the intention of the rule - that unrelated to aliases, you can't import anything outside of the current directory - including project code.

**Comment by jonathanj** [comment]
Okay, that's fair. I think I misinterpreted the documentation then.

So what I what then is actually a restricted import based on a pattern. Unfortunately it's not possible to specify per-pattern messages, which produces somewhat cryptic error messages but that's a different issue.

One suggestion then, I think the documentation for `no-relative-parent-imports` listing this explicitly in the "problem cases" example improve general interpretation.

**Comment by ljharb** [comment]
A docs PR would be very helpful!

**Comment by luxalpa** [comment]
As you can clearly see, you are the only one with this opinion and everyone else gets confused by it. Please just fix it.

Relative in case of a file path still refers to the URI (i.e. `../something` is always considered a relative file path). There is no such thing as a relative resolved path. You can compare resolved paths and then determine whether one resource is a parent of another, and you can create a relative file path to get from one resource to the other, but using the word `relative` redundantly here is rubbish. See also [all results on Google](https://www.google.com/search?q=relative+import). As far as I can see, this single rule is the only case on the whole internet where relative has a different (and therefore unexpected) meaning. I have not ever heard or read someone using `relative parent` to refer to anything other than the path.
If you disagree, then please explain, what is an `absolute parent import`?

**Comment by ljharb** [comment]
It's not an opinion, it's the way ES Modules work. Specifiers are NOT file paths there - in both node and browsers, they're URLs.

I understand you're saying that people see "relative" and assume it's applying to the URL in their source code, as opposed to the URL of the resolved file.

I have no idea what an "absolute parent import" is. Paths and URLs are absolute or relative (URLs can also be root-relative, with a leading slash).

The purpose of this rule would be violated if someone was able to get around it by using an alias.

**Comment by ljharb** [comment]
I agree "relative parent" is redundant.

"no-relative-parent-imports" is thus equivalent to "no-parent-imports", and "absolute" doesn't come into it at all.

The churn from renaming the rule is unlikely to ever be worth it.
